File: app/src/main/java/com/armaan/summarizer/SummaryActivity.java
package com.armaan.summarizer;

import android.content.Intent;
import android.content.SharedPreferences;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.armaan.summarizer.models.Summary;

import java.io.IOException;
import java.util.ArrayList;

public class SummaryActivity extends AppCompatActivity {

    TextView summaryText;
    TextView summaryTitle;
    Summary summary;
    SharedPreferences sharedPreferences;
    ArrayList<Summary> summaries;
    boolean favourite;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_summary);

        //Declaration//
        final Button button = (Button) findViewById(R.id.button);
        favourite = false;
        summaryText = (TextView) findViewById(R.id.summaryText);
        summaryTitle = (TextView) findViewById(R.id.summaryTitle);
        sharedPreferences = this.getSharedPreferences("com.armaan.summarizer",MODE_PRIVATE);
        Intent intent = getIntent();

        try {
            summary = (Summary) ObjectSerializer.deserialize(intent.getStringExtra("summaryText"));
            this.summaryText.setText(summary.getText());
            this.summaryTitle.setText(summary.getTitle());
            summaries = new ArrayList<Summary>();
            summaries = (ArrayList<Summary>) ObjectSerializer.deserialize(sharedPreferences.getString("summaries","kk"));

            button.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    try {
                        if(!favourite){
                            //summaries = new ArrayList<Summary>();
                            summaries.add(summary);
                            sharedPreferences.edit().putString("summaries",ObjectSerializer.serialize(summaries)).apply();
                            Toast.makeText(getApplicationContext(),"SummaryActivity added!",Toast.LENGTH_LONG).show();
                            button.setText("Remove");
                            favourite = true;
                        } else {
                            summaries.remove(summary);
                            sharedPreferences.edit().putString("summaries",ObjectSerializer.serialize(summaries)).apply();
                            Toast.makeText(getApplicationContext(),"SummaryActivity removed!",Toast.LENGTH_LONG).show();
                            button.setText("Add to Favourites");
                            favourite = false;
                        }
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
            });
        } catch (IOException e) {
            e.printStackTrace();
        }




    }

}

File: app/src/main/java/com/armaan/summarizer/SummaryActivity.kt
package com.armaan.summarizer

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Build
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.annotation.RequiresApi
import android.support.design.widget.FloatingActionButton
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.widget.*

import com.armaan.summarizer.models.Summary
import com.armaan.summarizer.summarizer.SummaryTool
import com.armaan.summarizer.utils.Utils
import com.squareup.picasso.Picasso
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import org.jsoup.Jsoup

import java.util.ArrayList

class SummaryActivity : AppCompatActivity() {

    internal lateinit var summaryText: TextView
    internal lateinit var summaryTitle: TextView
    internal lateinit var summaryImage: ImageView
    internal lateinit var progressBar: ProgressBar
    internal lateinit var fab: FloatingActionButton

    internal lateinit var summary: Summary
    internal lateinit var summaries: ArrayList<Summary>
    internal lateinit var sharedPreferences: SharedPreferences
    internal var favourite: Boolean = false
    internal lateinit var utils: Utils

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_summary)

        utils = Utils()

        summaryText = findViewById<View>(R.id.summaryText) as TextView
        summaryTitle = findViewById<View>(R.id.summaryTitle) as TextView
        summaryImage = findViewById<View>(R.id.summaryImage) as ImageView
        progressBar = findViewById(R.id.progressBar) as ProgressBar
        fab = findViewById(R.id.fab) as FloatingActionButton
        sharedPreferences = this.getSharedPreferences("com.armaan.summarizer", Context.MODE_PRIVATE)

        val intent = intent
        val action = intent.action
        val type = intent.type
        if (Intent.ACTION_SEND == action && type != null) {
            if ("text/plain" == type) {
                handleSendText(intent)
            }
        } else {
            generateSummary(intent.getStringExtra("summaryText"),intent.getStringExtra("image"))
        }

        favourite = false
        summaries = utils.deserialize(sharedPreferences.getString("summaries", ""))

        fab.setOnClickListener {
            if (!favourite ) {

                summaries!!.add(summary)
                sharedPreferences
                        .edit()
                        .putString("summaries", utils.serialize(summaries))
                        .apply()
                Toast.makeText(applicationContext, "Summary added!", Toast.LENGTH_LONG).show()
                fab.setImageDrawable(getDrawable(R.drawable.star))
                favourite = true
            } else {
                summaries.removeAt(remove(summary,summaries))
                sharedPreferences
                        .edit()
                        .putString("summaries", utils.serialize(summaries)).apply()
                Toast.makeText(applicationContext, "Summary removed!", Toast.LENGTH_LONG).show()
                fab.setImageDrawable(getDrawable(R.drawable.star_border))
                favourite = false
            }
        }
    }

    @SuppressLint("RestrictedApi")
    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    fun generateSummary(url: String, image: String?){
        doAsync {
            val doc = Jsoup.connect(url).get()
            val summaryTool = SummaryTool()
            summaryTool.init(doc.text())
            summaryTool.extractSentenceFromContext()
            summaryTool.groupSentencesIntoParagraphs()
            summaryTool.createIntersectionMatrix()
            summaryTool.createDictionary()
            summaryTool.createSummary()
            val text = summaryTool.summary.replace("[^\\x00-\\x7f]+".toRegex(), "")
            summary = Summary(doc.title(), text, url, image)

            uiThread {
                if(contains(summary,summaries) == true){
                    favourite = true
                    fab.setImageDrawable(getDrawable(R.drawable.star))
                } else{
                    favourite = false
                    fab.setImageDrawable(getDrawable(R.drawable.star_border))
                }
                fab.visibility = VISIBLE
                summaryText.text = summary.text
                summaryTitle.text = summary.title
                if(summary.imageUrl != null)
                    Picasso.get().load(summary.imageUrl).into(summaryImage);
                else
                    summaryImage.setVisibility(View.GONE)
                progressBar.visibility = GONE
            }
        }
    }

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    internal fun handleSendText(intent: Intent) {
        val sharedText = intent.getStringExtra(Intent.EXTRA_TEXT)
        if (sharedText != null) {
            generateSummary(sharedText,null)
        }
    }

    fun contains(summary: Summary,list: ArrayList<Summary>): Boolean {
        for (i in list){
            if (summary.title == i.title) return true
        }
        return false
    }

    fun remove(summary: Summary,list: ArrayList<Summary>): Int {
        for (i in 0..list.size){
            if (summary.title == list[i].title) return i
        }
        return -1
    }
}
